test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/service"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	gcecloud "k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	compute "google.golang.org/api/compute/v1"
+)
+
+const (
+	firewallTimeoutDefault  = 3 * time.Minute
+	firewallTestHttpTimeout = time.Duration(1 * time.Second)
+	// Set ports outside of 30000-32767, 80 and 8080 to avoid being whitelisted by the e2e cluster
+	firewallTestHttpPort = int32(29999)
+	firewallTestUdpPort  = int32(29998)
+)
+
+var _ = framework.KubeDescribe("Firewall rule", func() {
+	var firewall_test_name = "firewall-test"
+	f := framework.NewDefaultFramework(firewall_test_name)
+
+	// This test takes around 3 minutes to run
+	It("[slow] [serial] should create valid firewall rules for LoadBalancer type service", func() {
+		framework.SkipUnlessProviderIs("gce")
+		c := f.ClientSet
+		ns := f.Namespace.Name
+		cloudConfig := framework.TestContext.CloudConfig
+		gceCloud := framework.TestContext.CloudConfig.Provider.(*gcecloud.GCECloud)
+		// This source ranges is just used to examine we have the same things on LB firewall rules
+		firewallTestSourceRanges := []string{"0.0.0.0/1", "128.0.0.0/1"}
+
+		// OnlyLocal service is needed to examine which exact nodes the requests are being forward to by the Load Balancer on GCE
+		By("Creating a LoadBalancer type service with onlyLocal annotation")
+		serviceName := "firewall-test-loadbalancer"
+		jig := NewServiceTestJig(c, serviceName)
+		svc := createOnlyLocalLoadBalancServiceWithPort(jig, ns, serviceName, firewallTestHttpPort, firewallTestSourceRanges)
+		defer func() {
```

Comment:
you want to change the type to !LoadBalancer before deleting it, otherwise if this is the last test cluster-teardown will shutdown the master and leak your lb resources